extensions/amp-analytics/0.1/visibility-helper.js

```diff
@@ -0,0 +1,397 @@
+    const now = Date.now();
+
+    if (visibility > 0) {
+      this.firstSeenTime_ = this.firstSeenTime_ || now;
+      this.lastSeenTime_ = now;
+      // Consider it as load time visibility if this happens within 300ms of
+      // page load.
+      if (!this.loadTimeVisibility_ && (now - this.createdTime_) < 300) {
+        this.loadTimeVisibility_ = visibility;
+      }
+    }
+
+    const prevMatchesVisibility = this.matchesVisibility_;
+    const timeSinceLastUpdate =
+        this.lastVisibleUpdateTime_ ? now - this.lastVisibleUpdateTime_ : 0;
+    this.matchesVisibility_ = (
+        visibility > this.visiblePercentageMin_ &&
+        visibility <= this.visiblePercentageMax_);
+
+    if (this.matchesVisibility_) {
+      if (!prevMatchesVisibility) {
+        // The resource came into view: start counting.
+        dev().assert(!this.lastVisibleUpdateTime_);
+        this.fistVisibleTime_ = this.fistVisibleTime_ || now;
+      } else {
+        // Keep counting.
+        this.totalVisibleTime_ += timeSinceLastUpdate;
+        this.continuousTime_ += timeSinceLastUpdate;
+        this.maxContinuousVisibleTime_ =
+            Math.max(this.maxContinuousVisibleTime_, this.continuousTime_);
+      }
+      this.lastVisibleUpdateTime_ = now;
+      this.minVisiblePercentage_ =
+          this.minVisiblePercentage_ > 0 ?
+          Math.min(this.minVisiblePercentage_, visibility) :
+          visibility;
+      this.maxVisiblePercentage_ =
+          Math.max(this.maxVisiblePercentage_, visibility);
+      this.lastVisibleTime_ = now;
+    } else if (!this.matchesVisibility_ && prevMatchesVisibility) {
+      // The resource went out of view. Do final calculations and reset state.
+      dev().assert(this.lastVisibleUpdateTime_ > 0);
+
+      this.maxContinuousVisibleTime_ = Math.max(
+          this.maxContinuousVisibleTime_,
+          this.continuousTime_ + timeSinceLastUpdate);
+
+      // Reset for next visibility event.
+      this.lastVisibleUpdateTime_ = 0;
+      this.totalVisibleTime_ += timeSinceLastUpdate;
+      this.continuousTime_ = 0;  // Clear only after max is calculated above.
+      this.lastVisibleTime_ = now;
+    }
+
+    return this.matchesVisibility_ &&
+        (this.totalVisibleTime_ >= this.totalTimeMin_) &&
+        (this.totalVisibleTime_ <= this.totalTimeMax_) &&
+        (this.maxContinuousVisibleTime_ >= this.continuousTimeMin_) &&
+        (this.maxContinuousVisibleTime_ <= this.continuousTimeMax_);
+  }
+
+  /**
+   * Computes time, assuming the object is currently visible, that it'd take
+   * it to match all timing requirements.
+   * @return {time}
+   * @private
+   */
+  computeTimeToWait_() {
+    const waitForContinuousTime = Math.max(
+        this.continuousTimeMin_ - this.continuousTime_, 0);
+    const waitForTotalTime = Math.max(
+        this.totalTimeMin_ - this.totalVisibleTime_, 0);
+    const maxWaitTime = Math.max(waitForContinuousTime, waitForTotalTime);
+    return Math.min(
```

Comment:
To be honest, most of these code a direct refactor. Though looking at this, it appears that the code is trying to avoid zeros. I.e. minimal as long as it's not zero.